chain/boltdb/trimmed.go

```diff
@@ -0,0 +1,247 @@
+package boltdb
+
+import (
+	"context"
+	"io"
+	"path"
+	"sync"
+
+	bolt "go.etcd.io/bbolt"
+
+	"github.com/drand/drand/chain"
+	chainerrors "github.com/drand/drand/chain/errors"
+	"github.com/drand/drand/log"
+)
+
+// trimmedStore implements the Store interface using the kv storage boltdb (native
+// golang implementation). Internally, Beacons are stored as JSON-encoded in the
+// db file.
+type trimmedStore struct {
+	sync.Mutex
+	db *bolt.DB
+
+	log log.Logger
+
+	requiresPrevious bool
+}
+
+// newTrimmedStore returns a Store implementation using the boltdb storage engine.
+func newTrimmedStore(ctx context.Context, l log.Logger, folder string, opts *bolt.Options) (*trimmedStore, error) {
+	dbPath := path.Join(folder, BoltFileName)
+	db, err := bolt.Open(dbPath, BoltStoreOpenPerm, opts)
+	if err != nil {
+		return nil, err
+	}
+	// create the bucket already
+	err = db.Update(func(tx *bolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists(beaconBucket)
+		return err
+	})
+
+	return &trimmedStore{
+		log: l,
+		db:  db,
+
+		requiresPrevious: chain.PreviousRequiredFromContext(ctx),
+	}, err
+}
+
+// Len performs a big scan over the bucket and is _very_ slow - use sparingly!
+func (b *trimmedStore) Len(context.Context) (int, error) {
+	var length = 0
+	err := b.db.View(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(beaconBucket)
+		// this `.Stats()` call is the particularly expensive one!
+		length = bucket.Stats().KeyN
+		return nil
+	})
+	if err != nil {
+		b.log.Warnw("", "boltdb", "error getting length", "err", err)
+	}
+	return length, err
+}
+
+func (b *trimmedStore) Close(context.Context) error {
+	err := b.db.Close()
+	if err != nil {
+		b.log.Errorw("", "boltdb", "close", "err", err)
+	}
+	return err
+}
+
+// Put implements the Store interface. WARNING: It does NOT verify that this
+// beacon is not already saved in the database or not and will overwrite it.
+func (b *trimmedStore) Put(_ context.Context, beacon *chain.Beacon) error {
+	return b.db.Update(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(beaconBucket)
+		key := chain.RoundToBytes(beacon.Round)
+		err := bucket.Put(key, beacon.Signature)
+		if err != nil {
+			b.log.Errorw("storing beacon", "round", beacon.Round, "err", err)
+		}
+		return err
+	})
+}
+
+// Last returns the last beacon signature saved into the db
+func (b *trimmedStore) Last(context.Context) (*chain.Beacon, error) {
+	beacon := chain.Beacon{}
+	err := b.db.View(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(beaconBucket)
+		cursor := bucket.Cursor()
+		b, err := b.getCursorBeacon(bucket, cursor.Last)
+		if err != nil {
+			return err
+		}
+
+		beacon.Round = b.Round
+		beacon.Signature = b.Signature
+		beacon.PreviousSig = b.PreviousSig
+		return nil
+	})
+	return &beacon, err
+}
+
+// Get returns the beacon saved at this round
+func (b *trimmedStore) Get(_ context.Context, round uint64) (*chain.Beacon, error) {
+	beacon := chain.Beacon{}
```

Comment:
in the normal `Get` flow we're creating a beacon here and creating another one in the `getBeacon` method - perhaps one of them can be omitted to avoid the extra copying

Reply:
Also minor, but it might be nice to have the `getBeacon` func below this one just for mental proximity